tests/pytorch/test_geometry.py

```diff
@@ -109,7 +109,9 @@ def check_knn(g, x, start, end, k):
 
 @pytest.mark.parametrize('algorithm', ['bruteforce-blas', 'bruteforce', 'bruteforce-sharemem'])
 @pytest.mark.parametrize('dist', ['euclidean', 'cosine'])
-def test_knn_cuda(algorithm, dist):
+@pytest.mark.parametrize('exclude_self', [False, True])
+@pytest.mark.parametrize('output_batch', [False, True])
+def test_knn_cuda(algorithm, dist, exclude_self, output_batch):
```

Comment:
We should also add the same test (`exclude_self` and `output_batch`) for `test_knn_cpu`.  In particular I would like to see whether all the underlying kNN algorithms always return the center nodes themselves.
Also, could we add test for `algorithm == 'nn-descent'` (unless there's some reason why we don't test it, @hetong007 )?

Reply:
I can add the corresponding tests to `test_knn_cpu`.  I'd assume that the `'nn-descent'` algorithm isn't tested because the tests check the indices, and `'nn-descent'` isn't guaranteed to give exact results, though I've now changed the checking function to make checking indices optional, (for the coincident points test case I just added), so I could just have it skip checking indices if it's the approximate algorithm, and instead just check that each node has the correct number of edges, etc.  Would that make sense?

Reply:
I consolidated the `test_knn_cpu` and `test_knn_cuda` implementations, so that they all test the same cases now.

I also tried `'nn-descent'` and I confirmed that it doesn't always get the same indices as the other algorithms.  It also doesn't handle `k` being larger than `F.min(x_segs)` in the same way as the other algorithms, so even without index checking, some of the test cases fail, but I could reorganize the code a little so that it would be consistent in that regard, if it would be useful.  It seems a little out of the scope of the current change, though, so it might be better to change that behaviour in a separate PR, unless you'd like that changed sooner.

Reply:
Yes, `'nn-descent'` is a randomized algorithm that approximates the results.

> instead just check that each node has the correct number of edges, etc

This sounds good, and probably the only appropriate test for this algorithm.